cpp/src/io/parquet/parquet_reader_impl.hpp

```diff
@@ -66,6 +66,16 @@ class reader::Impl {
   table read(int skip_rows, int num_rows, int row_group);
 
  private:
+  /**
+   * @brief Align a size such that aligned loads within a memory block
+   * won't read bytes beyond the unaligned cuda-memcheck limit
+   *
+   * @param[in] size in bytes
+   *
+   * @return size_t Aligned size
+   **/
+  size_t align_size(size_t v) const { return v + (3 & -static_cast<int>(v)); }
```

Comment:
The arithmetic here looks a little bit magical. I'm not sure exactly how it works. Can you explain it?  You could write the `3` as `0x11` just to make it clear that it's a mask of the lower two bits.

Reply:
> The arithmetic here looks a little bit magical. I'm not sure exactly how it works. Can you explain it? You could write the `3` as `0x11` just to make it clear that it's a mask of the lower two bits.

The value we're adding to get to the next 4B-aligned boundary is
"(4 - size % 4) % 4", or (4 - (size & 3)) & 3, or just (-size & 3)
The static_cast is only here so that an overzealous compiler doesn't complain about the use of a negate operand on an unsigned type (I now prefer this over the '(size+3)&~3).
I don't think using 0b11 makes it clearer (perhaps I should expand to comment to make it clear that it's aligning the size to a 4B boundary and not say 8 or 64, but the '3' in the expression along with the name of the function 'align_size' isn't exactly obscure about what the function is doing).

Reply:
The compiler will do all of these optimizations. (compilers know to replace `%4` with `&3`, and they distribute operators), so I think the hand optimization obfuscates the code unnecessarily. I recommend just leaving it as `(4 - size % 4) % 4`, or at least `(-static_cast<int>(v) % 4)`.

All that said, a clearer thing to do would be to use cudf's existing safe rounding code, which eliminates potential overflows when rounding up. https://github.com/rapidsai/cudf/blob/d164d44200a7cae0852b18c2302c4842f13342fe/cpp/src/utilities/integer_utils.hpp#L33

Reply:
In general, I find it actually painful to use modulo power of 2, and also find that using modulo actually obfuscate the intent here, since it's fundamentally a logical masking out of bits that we want.

Reply:
Perhaps you would be more comfortable with (v + 3) & ~3 ?

Reply:
In the interest of not wasting any more time, I tried to use util::round_up_safe<size_t>(v, 4), which resulted in a "internal compiler error (codegen)", perhaps due to the mix of constexpr and std::throw ? It's got to be the most inefficient function to round up I have ever seen (though I suppose there is no limit to how inefficient something might be), and the sole reason for the existence of this function is the questionable design decision to use int as gdf_size_type, so I gave it my best shot, but I will respectfully stick to my original implementation.

Reply:
I agree the implementation that is here is unnecessarily obfuscated. 

> the questionable design decision to use int as gdf_size_type

Straight from the Arrow spec:
```
Array lengths
Any array has a known and fixed length, stored as a 32-bit signed integer, so a maximum of 2^31 - 1 elements. We choose a signed int32 for a couple reasons:

Enhance compatibility with Java and client languages which may have varying quality of support for unsigned integers.
To encourage developers to compose smaller arrays (each of which contains contiguous memory in its leaf nodes) to create larger array structures possibly exceeding 2^31 - 1 elements, as opposed to allocating very large contiguous memory blocks.
```

If perhaps you believe we should use an unsigned type, then the Google Style Guide has a good explanation on why we don't do this: https://google.github.io/styleguide/cppguide.html#Integer_Types

> perhaps due to the mix of constexpr and std::throw

You're allowed to `throw` in a `constexpr` so long as there exists arguments to the function that allow it to be evaluated at compile time without throwing: https://stackoverflow.com/questions/34280729/throw-in-constexpr-function


>  It's got to be the most inefficient function to round up I have ever seen

That's because it guards against overflow.

Reply:
It doesn't prevent uoverflow, it just throws an exception (and again overflow only likely because of the int32_t size type, which has nothing to do with what happens in this PR that uses size_t)
Back to the PR, using cudf::util::round_up_safe(v, 4) results in compile error because there is no template for <size_t, int> and using round_up_safe<size_t>(v, 4) results in internal compiler error (codegen), I suppose util::round_up_safe(v, static_cast<size_t>(4)) will work, but seriously ? As opposed to (v+3)&~3 or v+(3 & -v) ? Are people really that unfamiliar with 2's complement these days ?

Reply:
Regarding round_up_safe() the throw condition can't be evaluated at compile time unless both arguments to the function are also constexpr (v is not a constexpr in this PR).

Reply:
So far, I have been given 3 different suggestions that sound good but don't work to replace something that is literally 2-3 arithmetic instructions. Can I use that as grounds to dismiss a review and merge this trivial change that doesn't even change any functionality to begin with ?
(I get the feeling we're gonna have very long discussions for more serious PRs like the orc writer that are full of 2's complement stuff)

Reply:
Speaking of arrow, from bit_util.h:
constexpr uint64_t RoundUpToPowerOf2(uint64_t value, uint64_t factor) {
  // DCHECK(IsPowerOf2(factor));
  return (value + (factor - 1)) & ~(factor - 1);
}

Reply:
It has nothing to do with familiarity with the code or not. I just cringe at undocumented magic numbers in any context. I like clear and well-documented code. I like not duplicating functionality we have in our library in utility functions already. 

I agree that the rounding utilities in libcudf aren't great. But the solution is to fix them, not to avoid them. I would prefer that PRs contribute to improving the utilities rather than adding another one.

As for questionable design decisions, as Jake pointed out the use of int32 is a well-documented Arrow design decision. It surprised me at first, too, but we are sticking with it for Arrow compatibility.